File: easy/ASCII Art/Pascal/ASCII Art.pas
program Answer;
{$H+}
uses SysUtils, Classes, Math;

var
    L: Int32;
    H: Int32;
    T: String;
    asciiArt: array of String;
    result: array of String;
    i: Int32;
    j: Int32;
    index: Int32;
    startIdx: Int32;
    endIdx: Int32;
    c: Char;
    substr: String;
    finalResult: String;
begin
    readln(L);
    readln(H);
    readln(T);
    T := UpperCase(T);

    SetLength(asciiArt, H);
    for i := 0 to H - 1 do
    begin
        readln(asciiArt[i]);
    end;

    SetLength(result, H);
    for i := 0 to H - 1 do
        result[i] := '';

// Pascal strings use 1-based indexing, so the loop starts at 1.
for j := 1 to Length(T) do
begin
    c := T[j];
    if (c >= 'A') and (c <= 'Z') then
        index := Ord(c) - Ord('A')
    else
        index := 26;

    for i := 0 to H - 1 do
    begin
        // startIdx assumes 1-based indexing for Copy function in Pascal
        startIdx := index * L + 1;
        endIdx := startIdx + L - 1;
        substr := Copy(asciiArt[i], startIdx, L);
        result[i] := result[i] + substr;
    end;
end;

    finalResult := '';
    for i := 0 to H - 1 do
        finalResult := finalResult + result[i] + sLineBreak;
    writeln(finalResult);

flush(StdErr); flush(output); // DO NOT REMOVE
end.

File: easy/ASCII Art/Lua/ASCII Art.lua
L = tonumber(io.read())
H = tonumber(io.read())
T = io.read():upper()

asciiArt = {}
for i = 1, H do
    asciiArt[i] = io.read()
end

result = {}
for i = 1, H do
    result[i] = ""
end

for i = 1, #T do
    local c = T:sub(i, i)
    local index
    if c:match("%a") and c >= 'A' and c <= 'Z' then
        index = string.byte(c) - string.byte('A')
    else
        index = 26
    end

    for j = 1, H do
        local startIdx = index * L + 1
        local endIdx = startIdx + L - 1
        result[j] = result[j] .. asciiArt[j]:sub(startIdx, endIdx)
    end
end

print(table.concat(result, "\n"))
